signer/src/stacks_api.rs

```diff
@@ -0,0 +1,265 @@
+//! A module with structs that interact with the Stacks API.
+
+use std::time::Duration;
+
+use blockstack_lib::chainstate::nakamoto::NakamotoBlock;
+use blockstack_lib::codec::StacksMessageCodec;
+use blockstack_lib::types::chainstate::StacksBlockId;
+use config::Config;
+use config::Environment;
+use config::File;
+use futures::StreamExt;
+use serde::Deserialize;
+use serde::Deserializer;
+
+use crate::block_observer::StacksInteract;
+use crate::error::Error;
+
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// A deserializer for the url::Url type.
+fn url_deserializer<'de, D>(deserializer: D) -> Result<url::Url, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    String::deserialize(deserializer)?
+        .parse()
+        .map_err(serde::de::Error::custom)
+}
+
+/// A struct for the entries in the signers Config.toml (which is currently
+/// located in src/config/default.toml)
+#[derive(Debug, serde::Deserialize)]
+pub struct StacksSettings {
+    api: StacksApiSettings,
+    node: StacksNodeSettings,
+}
+
+/// Whatever
+#[derive(Debug, serde::Deserialize)]
+pub struct StacksApiSettings {
+    /// TODO(225): We'll want to support specifying multiple Stacks API
+    /// endpoints.
+    #[serde(deserialize_with = "url_deserializer")]
+    endpoint: url::Url,
+}
+
+/// Settings associated with the stacks node that this signer uses for information
+#[derive(Debug, serde::Deserialize)]
+pub struct StacksNodeSettings {
+    /// TODO(225): We'll want to support specifying multiple Stacks Nodes
+    /// endpoints.
+    #[serde(deserialize_with = "url_deserializer")]
+    endpoint: url::Url,
+}
+
+impl StacksSettings {
+    /// Create a new StacksSettings object by reading the relevant entries
+    /// in the signer's config.toml. The values there can be overridden by
+    /// environment variables.
+    ///
+    /// # Notes
+    ///
+    /// The relevant environment variables and the config entries that are
+    /// overridden are:
+    ///
+    /// * SIGNER_STACKS_API_ENDPOINT <-> stacks.api.endpoint
+    /// * SIGNER_STACKS_NODE_ENDPOINT <-> stacks.node.endpoint
+    ///
+    /// Each of these overrides an entry in the signer's `config.toml`
+    pub fn new_from_config() -> Result<Self, Error> {
+        let source = File::with_name("./src/config/default");
+        let env = Environment::with_prefix("SIGNER")
+            .prefix_separator("_")
+            .separator("_");
+
+        let conf = Config::builder()
+            .add_source(source)
+            .add_source(env)
+            .build()
+            .map_err(Error::SignerConfig)?;
+
+        conf.get::<StacksSettings>("stacks")
+            .map_err(Error::StacksApiConfig)
+    }
+}
```

Comment:
Can we move this to `config.rs` together with the rest of the configuration management?

Reply:
Oh right, of course. Will do.